text/0000-default-field-values.md

```diff
@@ -0,0 +1,1767 @@
+  | FieldsAndBase:{ fields:StructExprField+ % "," "," ".." base:Expr }
+  ;
+StructExprField = attrs:OuterAttr* kind:StructExprFieldKind;
+StructExprFieldKind =
+  | Shorthand:IDENT
+  | Explicit:{ field:FieldName ":" expr:Expr }
+  ;
+```
+
+the rule `StructExprFieldsAndBase` is extended with:
+
+```rust
+StructExprFieldsAndBase =| FieldsAndDefault:{ fields:StructExprField+ % "," "," ".." };
+```
+
+### Static semantics
+
+#### Defining defaults
+
+Given a `RecordField` where the default is specified, i.e.:
+
+```rust
+RecordField = attrs:OuterAttr* vis:Vis? name:IDENT ":" ty:Type "=" def:Expr;
+```
+
+all the following rules apply when type-checking:
+
+1. The expression `def` must be a constant expression.
+
+2. The expression `def` must unify with the type `ty`.
+
+When lints check attributes such as `#[allow(lint_name)]` are placed on a
+`RecordField`, it also applies to `def` if it exists.
+
+#### Initialization expressions
+
+Given an expression `τ` of the form (e.g. `Foo { x: 1, .. }`):
+
+```rust
+path:Path "{" attrs:InnerAttr* fields:StructExprField+ % "," "," ".." "}"
+```
+
+all the following rules apply when type-checking:
+
+1. All `fields` in `τ` mentioned must exist in the `struct` or `enum` variant
+   referenced by `path`. No field in `fields` may be mentioned twice.
+
+2. For each `field: expr` pair in `fields` of `τ`,
+   `expr` must unify with the type of `field` in the `struct`
+   or `enum` variant referenced by `path`.
+
+3. For each `field: expression` pair in `fields` of `τ`,
+   `field` must be visible in the context which `τ` occurs in.
+
+4. For each field `f` in `τ` *not* mentioned in `fields`, a default value must
+   exist for `f` in the `struct` or `enum` variant referenced by `path`.
+
+---------------------
+
+**Lemma 1:** If `τ` is of a type `σ` where `σ: Sync`, and if all `field: expr`
+has an `expr` component which is a valid constant expression, then `τ` is
+also a valid constant expression.
+
+**Proof:** It trivially holds that if all `expr`s in `τ` are
+constant expressions and all default values for fields omitted
+are also constant expressions, then `τ` is a constant expression.
+
+**Lemma 2:** If an expression `x` of form `Path { .. }` is of a type `σ` where
+`σ: Sync`, then `x` is a constant expression.
+
+**Proof:** `x` is a special case of `τ` with all fields according to the default
+values. Since there are no explicitly provided fields, by Lemma 1 it holds.
+
+### Dynamic semantics
+
+Given a well-formed expression `τ` according to the aforementioned form,
+the operational semantics is the same as the semantics of an expression
+`path { new_fields }` with `new_fields = fields ∪ default_fields` where:
+
++ `∪` computes the left-biased union where equal elements are determined
+  solely on equality of field names.
+
++ `fields` are the set of explicitly given `field: expr` pairs.
+
++ `default_fields` are the fields in `τ`'s data-type or variant which
+  have provided default values.
```

Comment:
I also think section this can be cut from the RFC. I don't think we go into this much detail for new expressions, and it feels very very verbose to explain their dynamic semantics, especially because the only important part that users *would* care about is something like initialization order which wouldn't differ from FRU anyways.